apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1099,34 +1119,105 @@ private async Task ProcessPullMessagesAsync(List<ReceivedMessage> msgs, HashSet<
+                    }
+                }
+
+                public void Cancel()
+                {
+                    CancellationTokenSource cts2;
+                    lock (_lock)
+                    {
+                        cts2 = _cts;
+                    }
+                    cts2?.Cancel();
+                }
+            }
+
+            private void HandleExtendLease(HashSet<string> msgIds, LeaseCancellation cancellation)
             {
                 if (_softStopCts.IsCancellationRequested)
                 {
                     // No further lease extensions once stop is requested.
                     return;
                 }
-                bool anyMsgIds;
-                lock (msgIds)
+                if (cancellation == null)
                 {
-                    anyMsgIds = msgIds.Count > 0;
-                    if (anyMsgIds)
+                    // Create a task to cancel lease-extension once max-lease-extension time has been reached.
+                    cancellation = new LeaseCancellation(_softStopCts);
+                    Add(_scheduler.Delay(_maxExtensionDuration, cancellation.Token), Next(false, () =>
                     {
-                        lock (_lock)
+                        cancellation.Dispose();
+                        lock (msgIds)
                         {
-                            _extendQueue.Enqueue(msgIds.Select(x => new TimedId(_extendThrottleHigh + 1, x)));
+                            msgIds.Clear();
                         }
-                    }
+                    }));
                 }
-                if (anyMsgIds)
+                if (!cancellation.IsDisposed)
                 {
-                    // Ids have been added to _extendQueue, so trigger a push.
-                    _eventPush.Set();
-                    // Some ids still exist, schedule another extension.
-                    Add(_scheduler.Delay(_autoExtendInterval, _softStopCts.Token), Next(false, () => HandleExtendLease(msgIds)));
-                    // Increment _extendThrottles.
-                    _extendThrottleHigh += 1;
-                    Add(_scheduler.Delay(_extendQueueThrottleInterval, _softStopCts.Token), Next(false, () => _extendThrottleLow += 1));
+                    bool anyMsgIds;
+                    lock (msgIds)
+                    {
+                        anyMsgIds = msgIds.Count > 0;
+                        if (anyMsgIds)
+                        {
+                            lock (_lock)
+                            {
+                                _extendQueue.Enqueue(msgIds.Select(x => new TimedId(_extendThrottleHigh + 1, x)));
+                            }
+                        }
+                    }
+                    if (anyMsgIds)
+                    {
+                        // Ids have been added to _extendQueue, so trigger a push.
+                        _eventPush.Set();
+                        // Some ids still exist, schedule another extension.
```

Comment:
Possibly edit this comment to add "with the same deadline for how long the lease can be extended"?